### PR TITLE
Only use jump address as fallback if host is unreachable

### PR DESCRIPTION
**Pull request by aoskotsky-amplify**
Change `disco_remote_exec.py` to use the jump address only if the host is not directly reachable. We are doing the same thing in `disco_ssh.py`.

The motivation for this is that `disco_aws.py exec` does not work from the deployenators at the moment. We are always trying to use the jump address when one is available but the deployenators in deploy/tunnel can't talk to staging/maintenance. However, they can directly connect to staging/intranet so skipping the jump box fixes the problem.

Future Work:
Maybe we should consider reusing some of the code in `disco_ssh.py` in `disco_remote_exec.py` since a lot of the code for building up the ssh command is similar.

@jchung-amplify @IlyaSukhanov @pchand-amplify

### CODE REVIEW

**Comment by aoskotsky-amplify** [comment]
build

**Comment by bsuzukiamplify** [comment]
Note: the [story](https://amplify-litco.atlassian.net/browse/PS-782) to create a Deployenator job to run tests is blocked on this.

File: disco_aws_automation/disco_remote_exec.py [hunk 1]
```diff
@@ -81,10 +82,18 @@ def remotecmd(address, remote_command, user, stdin=None,
                   ssh_options=(), forward_agent=None):
         """
         Runs the passed in command on a remote host, via a jump host if a jump_address
-        is provided.
+        is provided and the address is not reachable without a jump host.
```
Comment (pchand-amplify) [inline disco_aws_automation/disco_remote_exec.py]: `and the address` to `and if the address`

File: disco_aws_automation/disco_remote_exec.py [hunk 2]
```diff
@@ -171,6 +180,19 @@ def _call_rsync(command):
...
                 return subprocess.call(command, stdout=devnull, stderr=devnull)
 
+    @staticmethod
+    def _is_reachable(ip_address):
+        """Returns True if we can connect to port 22 at the given ip address"""
+        if not ip_address:
+            return False
+        logging.info("Probing %s", ip_address)
+        try:
+            sock = socket.create_connection((ip_address, 22), timeout=2)
```
Comment (pchand-amplify) [inline disco_aws_automation/disco_remote_exec.py]: Should we use constant variables to represent `22` and `2`?

**Comment by pchand-amplify** [comment]
👍  I would have liked to see the refactoring done to make sure we use the common code to ssh in `disco_ssh.py` and `disco_remote_exec.py`, but it can be a story by itself.

**Comment by IlyaSukhanov** [comment]
+1

**Comment by jconstance-amplify** [comment]
build

**Pull request merged**